Add post-process functionality in write_with_gap

rish9101 opened the pull request:
We had ported post-process to custom_mutators, but write_with_gap function was still missing the important post-process usage. This commit adds that as raised in #457 .

rish9101 committed:
Add post-process functionality in write_with_gap

vanhauser-thc commented:
Hey rishi, thanks for the PR ... it looks wrong on so many levels :)
first you seem to always copy the buffer even if there is no custom trimmer. second the copy buffer is not using ck_maybe_alloc (or what that function was called) so it is costly every single trim attempt. lastly - why not update the length after the custom postprocess? and fail if it is a bad value?

rish9101 commented:
Ohh, sorry about it. I'll correct the redundant copying part. Since ck_alloc (or any allocation) may be costly, I think we can use a local buffer for this? And the `new_size` is updated after every `post_process` function in the list, I do need to throw a FATAL though.

vanhauser-thc commented:
hey @rish9101 no hurry, just want to let you know: once this PR is done we will push again from dev to stable :) this PR should be in there

rish9101 committed:
Remove reduntant copying from write_with_gap function

rish9101 commented:
Hey @vanhauser-thc, sorry for the delay. I've removed the redundant copying and added error handling for `new_size` and `new_buf` now :) . Also, I've used a local buffer for the temporary copy of the trimmed data (if it requires making a copy for post-proessing).

vanhauser-thc commented on src/afl-fuzz-run.c:
```diff
@@ -142,18 +142,82 @@ static void write_with_gap(afl_state_t *afl, void *mem, u32 len, u32 skip_at,
+  u8 mem_trimmed[skip_at + tail_len +
+                 1];  // 1 extra size to remove chance of overflow
+
+  ssize_t new_size = skip_at + tail_len;
```
can you please switch this to `new_size = len - skip_len`? that makes it easier to follow :) also for `mem_trimmed` above

vanhauser-thc commented on src/afl-fuzz-run.c:
```diff
@@ -142,18 +142,82 @@ static void write_with_gap(afl_state_t *afl, void *mem, u32 len, u32 skip_at,
+
+        if (post_process_skipped) {
+
+          if (skip_at) { memcpy(mem_trimmed, (u8 *)mem, skip_at); }
```
why not move this just before the FOREACH loop? if there is at least 1 mutator to run, this has to be done, but it does not need to be checked for every entry in the loop.

rish9101 replied:
This has an advantage that we only copy the trimmed data into a new buffer if we actually have some post processing to do though. And that I think should be better than the cost of a bool check maybe?

vanhauser-thc replied:
I think you are right, leave it like this!

vanhauser-thc commented on src/afl-fuzz-run.c:
```diff
@@ -142,18 +142,82 @@ static void write_with_gap(afl_state_t *afl, void *mem, u32 len, u32 skip_at,
+  */
+  u8 mem_trimmed[skip_at + tail_len +
+                 1];  // 1 extra size to remove chance of overflow
+
```
this is something you cannot know - trimming (and post processing) of a custom module can increase the size (e.g. for grammar, a simpler grammar can make the file size larger). I dont know what a good safe value is, but maybe add 4096 instead of 1. and please make a comment that this can break.

rish9101 replied:
We can have extra buffer. But, `write_with_gap` actually is called only when AFL's own trimming function is called, so I don't think they can ever exceed the length given here. Also, `post_process` is supposed to return the processed data in its own buffer. So, the length of this buffer I think should be fine :)

vanhauser-thc replied:
again, I think you are correct. nothing to see here :)

domenukk commented on src/afl-fuzz-run.c:
```diff
@@ -142,18 +142,82 @@ static void write_with_gap(afl_state_t *afl, void *mem, u32 len, u32 skip_at,
+  This memory is used to carry out the post_processing(if present) after copying
+  the testcase by removing the gaps
+  */
+  u8 mem_trimmed[skip_at + tail_len +
```
So far, we tried to avoid the use of variable length arrays, as they are error-prone (mainly, the can be too large for the stack). I think we may want to reuse one of the alrady allocated buffers in `alf_state_t` or allocate a new one (?)

domenukk replied:
Alternatively, the array could be fixed size to the maximum testcase size, that feels even cleaner.

rish9101 replied:
I didn't know about the allocated buffers in `afl_state_t` actually. 😅

vanhauser-thc replied:
You can safely do  `u8 *mem_trimmed = ck_maybe_grow(BUF_PARAMS(trim), len - skip_len + 1);`  here with no need to free or do anything else. After that the PR is read to land :)

Commit:
Minor change to write_with_gap

vanhauser-thc commented:
sorry this takes too long, doing it myself now ... ;)
